lib-impl/src/main/java/org/ergoplatform/appkit/impl/UnsignedTransactionBuilderImpl.scala

```diff
@@ -65,36 +68,46 @@ class UnsignedTransactionBuilderImpl(val _ctx: BlockchainContextImpl) extends Un
   }
 
   override def tokensToBurn(tokens: ErgoToken*): UnsignedTransactionBuilder = {
-    Collections.addAll(_tokensToBurn, tokens:_*)
+    require(_tokensToBurn.isEmpty, "Tokens to burn already specified.")
+    _tokensToBurn = Some({
+      val res = new util.ArrayList[ErgoToken]()
+      Collections.addAll(res, tokens:_*)
+      res
+    })
     this
   }
 
-  private def appendOutputs(outputs: OutBox*): Unit = {
-    val boxes = outputs.map(c => c.asInstanceOf[OutBoxImpl].getErgoBoxCandidate)
-    Collections.addAll(_outputCandidates, boxes:_*)
-  }
-
   override def sendChangeTo(changeAddress: ErgoAddress): UnsignedTransactionBuilder = {
-    require(_changeAddress == null, "Change address is already specified")
-    _changeAddress = changeAddress
+    require(_changeAddress.isEmpty, "Change address is already specified")
+    _changeAddress = Some(changeAddress)
     this
   }
 
+  def checkNonEmpty[T](list: Option[List[T]], msg: => String) = {
+    require(list.isDefined && !list.get.isEmpty, msg)
+  }
+
   override def build: UnsignedTransaction = {
-    val boxesToSpend = _inputBoxes
+    checkNonEmpty(_inputBoxes, "Input boxes are not specified")
+    checkNonEmpty(_outputCandidates, "Output boxes are not specified")
+    val boxesToSpend = _inputBoxes.get
       .map(b => ExtendedInputBox(b.getErgoBox, b.getExtension))
-    val dataInputBoxes = _dataInputBoxes.map(b => b.getErgoBox)
+    val dataInputBoxes = _dataInputBoxes
+       .getOrElse(new util.ArrayList[InputBoxImpl]())
+       .map(b => b.getErgoBox)
     val dataInputs = JavaHelpers.toIndexedSeq(_dataInputs)
     require(_feeAmount.isEmpty || _feeAmount.get >= MinFee,
       s"When fee amount is defined it should be >= $MinFee, got ${_feeAmount.get}")
-    require(_changeAddress != null, "Change address is not defined")
-    val outputCandidates = JavaHelpers.toIndexedSeq(_outputCandidates)
+    require(_changeAddress.isDefined, "Change address is not defined")
+    val outputCandidates = JavaHelpers.toIndexedSeq(_outputCandidates.get)
```

Comment:
Would be good to remove get as its considered anti-pattern

Reply:
fixed